### ISSUE TITLE
Adding @discardableResult to `then`

### ISSUE THREAD

**Issue by FilipZawada** [mxcl/PromiseKit#595]
I'd expect promises to return @discardableResult, to avoid compiler warning.

Right now this:
```
promise.then {
  ...
}.then {
  ...
} 
```

Causes compiler to warn about unused result. To prevent you have to do:

```
_ = promise.then {
  ...
}.then {
  ...
} 
```

@mxcl What about me making a PR for this?

**Comment by zlangley** [comment]
Dupes #549, #551, #561, #574, #584.

**Comment by FilipZawada** [comment]
Ah sorry :-) At least it'd be work to re-work tests, so they don't spam with compiler warnings so much.

**Comment by mxcl** [comment]
I think we need to some work here at least, eg. I use `after(interval: _)` a lot, there's no reason to warn for that. It cannot fail.

So maybe we should add an `UnfailablePromise` class.

Also, as an aside, I added a `fatalCatch()` chain terminator to my local branch, and use that quite a lot.

**Comment by mxcl** [comment]
Also, during my use of PromiseKit, almost always I find that if I am getting this warning it means I am not handling errors. Only occasionally am I added a sub-chain where I know I am handling the error elsewhere and thus adding a `_ =` prefix is necessary and thus also a good indicator that I know the error is handled elsewhere.

Other than the above mentioned cases where the promise cannot fail.

Another thing on this, this warning almost makes the `UnhandledErrorHandler` unnecessary, since the Swift compiler is ensuring we handle errors. We may want to consider deprecating it, unless I'm missing something.

**Comment by mxcl** [comment]
The other option of course is to add the `@discardableResult` annotation.

I feel this would lead to less good code for our users.

But it would lead to less tedious usage for our users.

Honestly, I feel how we have it *now* is more Swifty. However, we are not Apple, and our popularity to an extent depends on us not being tedious.

**Comment by zlangley** [comment]
Since this is consistently brought up, we might want to add something about it to the FAQ. We could also mention your `fatalCatch()` suggestion. (Although I'm not sure people are checking the FAQ before creating issues here.)

I agree that something like `UnfailablePromise` could simplify a several APIs (`after`, non-throwing `recover`, non-throwing `UnfailablePromise.then`) and would therefore be worthwhile.

It would be nice if we could customize the unused result warning to indicate that you're expected to terminate promises with a catch, but I don't think that's currently possible.

**Comment by FilipZawada** [comment]
I like the idea of `UnfailablePromise`. I think we can make PromiseKit smart enough, that it analyzes whole chain and decides whether it may fail or not. It would follow simple rules when chaining:
```text
failable + failable -> failable
unfailable + failable -> failable
failable + unfailable -> failable
unfailable + unfailable -> unfailable
```

The promise would become unfailable if:
* promise body __not__ `throws`
* was constructed using an unfailable initializer like `Promise { success: ()->Void in }`
* promise body returned an unfailable promise

So this chain would be unfailable:
```swift
func measureUserProgress(user: User, timespan: TimeInterval) -> 
    firstly { () -> Int           // user.getScore() doesn't throw, so this promise can't fail
        return user.getScore()
    }.then { score in             // after is unfailable; then {score} is unfailable too
       return after(timespan).then { score }
    }.then { oldScore in          // since user.getScore() doesn't throw, this would be unfailable too
       let newScore = user.getScore()
       return newScore - oldScore
    }
}

measureUserProgress(user).then { progress in
    print("progress: \(progress)")
} 
```

**Comment by mxcl** [comment]
Can we do all these smarts so it is compile time? I didn't think so, but would love to be wrong.

**Comment by FilipZawada** [comment]
I don't think either (although haven't tried that hard yet).

If there would be some simple logic - is it a deal breaker?

**Comment by FilipZawada** [comment]
Now I think that's doable. I'll try to create a PoC once I got more time.

**Comment by FilipZawada** [comment]
That's how we can do it:
```
class UnfailablePromise<T> {
    @discardableResult
    func then() -> UnfailablePromise<T> { fatalError("we want generic protocols!") }
}

class Promise<T>: UnfailablePromise<T> {
    override func then() -> Promise<T> {
        return Promise<T>()
    }
}

let promise: UnfailablePromise<Int> = Promise<Int>()
promise.then() // no warning

let promise2: Promise<Int> = Promise()
promise2.then() // warn: unused result
```

However it makes use of feature that is considered a swift 3 [bug](https://bugs.swift.org/browse/SR-2198).

Regarding swift bug, we can:
a) assume that before swift team fixes that bugs, they will give us much needed generic protocols
b) maybe we could do other way around, so `UnfailablePromise` inherits from Promise. I'll have a think tomorrow.

**Comment by FilipZawada** [comment]
I don't really think `@discardableResult` should be inherited. Therefore I made a [comment](https://bugs.swift.org/browse/SR-3377) to open that discussion. Once I get some feedback from swift team, I can make a PR here.

**Comment by FilipZawada** [comment]
I made a significant progress today ([branch](https://github.com/FilipZawada/PromiseKit/tree/595-discardable-warning)). Snippet below works very well:

```swift
        // unfailable + unfailable #1
        Promise<Int>.with(5).then { val in
            return Promise(value: val + 10)
        }

        // unfailable + unfailable #2
        Promise<Int>.with(5).then { (val: Int) -> Int in
            return val + 10
        }

        // unfailable + failable #1 -> warning
        _ = Promise<Int>.with(5).then { val throws -> Int in
            return val + 5
        }

        // failable + unfailable -> warning
        let p = Promise<Int>.with(E.dummy).then { val -> Int in
            return val + 5
        }
        p.catch { _ in }

        // failable + failable -> warning
        _ = Promise(value: 5).then { val throws -> Int in
            return val + 5
        }
```

However there are still many unknown parts, so I don't know if this is 100% doable yet.

@mxcl can you please review existing solution to see if it has a potential to get into production? I just don't want to do unnecessary work.

**Comment by mxcl** [comment]
Great stuff.

I haven't looked yet, but I spent my time in transit today trying out a PromiseKit where Promise is based on a protocol `Thennable`, this would presumably make it easy for UnfailablePromise to a separate protocol and not need to inherit.

Just saying. I'm tired now and can't recall all the reasons I have done this. Will likely push the work to a branch soon so you guys can take a look. Certainly, the code is simpler in many places.

Anyway, will check out your work @FilipZawada thanks, I'm super hyped this is possible.

**Comment by mxcl** [comment]
The `experimental-5.x` is now beta worthy IMO, I'll be using it in an app from today.

**Comment by FilipZawada** [comment]
Awesome! I'll give it a try in my side project

**Comment by mxcl** [comment]
There *will* be frequent updates, also please, your feedback would be wonderful.

**Comment by mohamede1945** [comment]
What about a `suppress` error function?
It's very similar to the `_ =` solution but I think it's more elegant as it clearly states its purpose.

**Comment by mxcl** [comment]
In PMK5 (experimental) I added a `cauterize()` function, that prints any error and terminates the chain. So yeah, would be ok with this being in PMK4.

**Comment by mohamede1945** [comment]
It looks very good. Is it possible to implement it with a tag parameter for the print? something like that:
```swift
public func cauterize(tag: String? = nil) {
    `catch` { error in
        if let tag = tag {
            print("PromiseKit: [\(tag)] unhandled error:", error)
        } else {
            print("PromiseKit: unhandled error:", error)
        }
    }
}
```

**Comment by mxcl** [comment]
Closed by https://github.com/mxcl/PromiseKit/commit/a5d5221684061b482111aaed5d7ca8a94bcd4f7b

Please read the commit for details.

TL;DR: I added `@discardableResult` for PMK 4. PMK 5 has been released and solves this more elegantly.
